compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -49,3 +97,1385 @@ fn variant_has_complex_ctor(variant: &ty::VariantDef) -> bool {
+        def.variants().iter().for_each(|variant| {
+            let (nonex_enum, nonex_var) = flag_non_exhaustive_variant(non_local_def, variant);
+            nonexhaustive_flag |= nonex_enum;
+            nonexhaustive_variant_flag |= nonex_var;
+        });
+
+        if nonexhaustive_flag {
+            FfiResult::new_with_reason(ty, fluent::lint_improper_ctypes_non_exhaustive, None)
+        } else if nonexhaustive_variant_flag {
+            FfiResult::new_with_reason(
+                ty,
+                fluent::lint_improper_ctypes_non_exhaustive_variant,
+                None,
+            )
+        } else {
+            def.variants()
+                .iter()
+                .map(|variant| {
+                    self.visit_variant_fields(state, ty, def, variant, args)
+                        // FIXME: check that enums allow any (up to all) variants to be phantoms?
+                        // (previous code says no, but I don't know why? the problem with phantoms is that they're ZSTs, right?)
+                        .forbid_phantom()
+                })
+                .reduce(|r1, r2| r1 + r2)
+                .unwrap() // always at least one variant if we hit this branch
+        }
+    }
+
+    /// Checks if the given type is "ffi-safe" (has a stable, well-defined
+    /// representation which can be exported to C code).
+    fn visit_type(
+        &mut self,
+        state: CTypesVisitorState,
+        outer_ty: Option<Ty<'tcx>>,
+        ty: Ty<'tcx>,
+    ) -> FfiResult<'tcx> {
+        use FfiResult::*;
+
+        let tcx = self.cx.tcx;
+
+        // Protect against infinite recursion, for example
+        // `struct S(*mut S);`.
+        // FIXME: A recursion limit is necessary as well, for irregular
+        // recursive types.
+        if !self.cache.insert(ty) {
+            return FfiSafe;
+        }
+
+        match *ty.kind() {
+            ty::Adt(def, args) => {
+                if let Some(inner_ty) = ty.boxed_ty() {
+                    return self.visit_indirection(
+                        state,
+                        outer_ty,
+                        ty,
+                        inner_ty,
+                        IndirectionType::Box,
+                    );
+                }
+                if def.is_phantom_data() {
+                    return FfiPhantom(ty);
+                }
+                match def.adt_kind() {
+                    AdtKind::Struct | AdtKind::Union => {
+                        // I thought CStr (not CString) could not be reached here:
+                        //   - not using an indirection would cause a compile error prior to this lint
+                        //   - and using one would cause the lint to catch on the indirection before reaching its pointee
+                        // but for some reason one can just go and write function *pointers* like that:
+                        // `type Foo = extern "C" fn(::std::ffi::CStr);`
```

Comment:
1. Because unsized function *parameters* are something we may want to support.
2. The code may not be well-formed: as you may have noticed at some point, you get warnings even if you get errors (usually), and this is because we lint even on "bad" code. This is because rustc didn't use to, once upon a time, and it was a bad debugging experience.